javalib/src/main/scala/java/lang/process/PipeIO.scala

```diff
@@ -31,60 +134,100 @@ private[lang] object PipeIO {
         ioStream.nullStream
     }
   }
+
   trait Stream extends InputStream {
     def process: GenericProcess
     def drain(): Unit = {}
   }
+
   class StreamImpl(val process: GenericProcess, is: FileInputStream)
-      extends BufferedInputStream(is)
-      with Stream {
-    override def available() = try {
-      super.available() match {
-        // Check the FileInputStream in case the BufferedInputStream hasn't been filled yet.
-        case 0 =>
-          in.available() match {
-            case 0 if !drained => availableFD()
-            case a             => a
-          }
-        case a => a
+      extends Stream {
+
+    private var src: InputStream = is
+
+    // By convention, caller is synchronized on 'this'.
+    private def availableUnSync() = {
+      src match {
+        case fis: FileInputStream => availableFD()
+        case _                    => src.available()
       }
-    } finally process.checkResult()
+    }
 
-    override def read(): Int = synchronized {
-      // Retry read if failed due to termination of the process (file descriptor would become invalidated)
-      // In such case output would be drained and underlying input stream would be replaced with fully read content
-      try super.read()
-      catch { case _: IOException if drained => super.read() }
+    override def available(): Int = synchronized {
+      try
+        availableUnSync()
       finally process.checkResult()
     }
-    override def read(buf: Array[scala.Byte], offset: Int, len: Int) =
-      try super.read(buf, offset, len)
-      catch { case _: IOException if drained => super.read(buf, offset, len) }
+
+    override def read(): Int = synchronized {
+      try
+        src.read()
       finally process.checkResult()
+    }
 
-    override def drain() = synchronized {
-      @tailrec
-      def loop(readBuf: Array[scala.Byte]): Array[Byte] = {
-        if (drained) readBuf
+    override def read(buf: Array[scala.Byte], offset: Int, len: Int): Int =
+      synchronized {
+
+        if (offset < 0 || len < 0 || len > buf.length - offset) {
+          val prefix =
+            s"Range [${offset}, ${offset} + ${len})"
+          val suffix =
+            s" out of bounds for length ${buf.length}"
+          throw new IndexOutOfBoundsException(s"${prefix}${suffix}")
```

Comment:
If there's nothing to prefix or suffix but rather concatenate we should do it one go. This now requires creating 3 String Builder instances. Instead we can use just a single one.
```suggestion
          throw new IndexOutOfBoundsException(
            s"Range [${offset}, ${offset} + ${len}) out of bounds for length ${buf.length}"
          )
```

Reply:
Thank you for the suggestion & for taking the "teachable moment".

It is an example of  how I should think about performance in (almost) every detail, and not just 
correctness. This case is on a "should never happen" error path, but I probably am prone to
using the same idiom in code where performance is more critical.

I had originally written that section in the more "natural" way you suggest.
I found that, with indentation, the message exceeded 80 characters of my,
and many others, terminal making it hard to read (or print to dead-trees).

I believe I have come up with a riff of you suggestion which solves both
the essential concern of the suggestion and my dislike of having to use 132
character terminals.  

If you do not like my edit, please call out and I can try something else.

Reply:
The additional benefit is that the change gave an additional run through CI, especially Windows.

Always good when in is, on the side, chasing intermittent failures.